### PR TITLE
Codegen const panic messages as function calls

### PR DESCRIPTION
**Pull request by Mark-Simulacrum**
This skips emitting extra arguments at every callsite (of which there
can be many). For a librustc_driver build with overflow checks enabled,
this cuts 0.7MB from the resulting shared library (see [perf]).

A sample improvement from nightly:

```
        leaq    str.0(%rip), %rdi
        leaq    .Lalloc_d6aeb8e2aa19de39a7f0e861c998af13(%rip), %rdx
        movl    $25, %esi
        callq   *_ZN4core9panicking5panic17h17cabb89c5bcc999E@GOTPCREL(%rip)
```

to this PR:

```
        leaq    .Lalloc_d6aeb8e2aa19de39a7f0e861c998af13(%rip), %rdi
        callq   *_RNvNtNtCsduqIKoij8JB_4core9panicking11panic_const23panic_const_div_by_zero@GOTPCREL(%rip)
```

[perf]: https://perf.rust-lang.org/compare.html?start=a7e4de13c1785819f4d61da41f6704ed69d5f203&end=64fbb4f0b2d621ff46d559d1e9f5ad89a8d7789b&stat=instructions:u

### CODE REVIEW

File: compiler/rustc_codegen_ssa/src/mir/block.rs [hunk 1]
```diff
@@ -651,10 +651,8 @@ impl<'a, 'tcx, Bx: BuilderMethods<'a, 'tcx>> FunctionCx<'a, 'tcx, Bx> {
                 (LangItem::PanicMisalignedPointerDereference, vec![required, found, location])
             }
             _ => {
-                let msg = bx.const_str(msg.description());
-                // It's `pub fn panic(expr: &str)`, with the wide reference being passed
-                // as two arguments, and `#[track_caller]` adds an implicit third argument.
-                (LangItem::Panic, vec![msg.0, msg.1, location])
+                // It's `pub fn panic_...()` and `#[track_caller]` adds an implicit argument.
+                (msg.description(), vec![location])
```
Comment (bjorn3) [inline compiler/rustc_codegen_ssa/src/mir/block.rs:655]: Is `#[lang = "panic"]` still necessary with this?

Comment (Mark-Simulacrum) [reply compiler/rustc_codegen_ssa/src/mir/block.rs:655]: I suspect no, I'll look at cleaning it up. I'm not particularly happy with the ~10+ lang items being added here though, mostly trying to get something fully working for perf.

**Comment by bjorn3** [comment]
The above CI failure is because https://github.com/rust-lang/rust/blob/master/compiler/rustc_codegen_cranelift/example/mini_core.rs and https://github.com/rust-lang/rust/blob/master/compiler/rustc_codegen_gcc/example/mini_core.rs also need the new lang items.

**Comment by RalfJung** [comment]
> The above CI failure is because https://github.com/rust-lang/rust/blob/master/compiler/rustc_codegen_cranelift/example/mini_core.rs and https://github.com/rust-lang/rust/blob/master/compiler/rustc_codegen_gcc/example/mini_core.rs also need the new lang items.

Any way we can avoid all those mini_core copies? I've had to update them a couple times as well, it is quite annoying.

File: library/core/src/panicking.rs [hunk 2]
```diff
@@ -145,6 +145,57 @@ pub const fn panic(expr: &'static str) -> ! {
...
+                //
+                // never inline unless panic_immediate_abort to avoid code
+                // bloat at the call sites as much as possible
+                #[cfg_attr(not(feature = "panic_immediate_abort"), inline(never), cold)]
+                #[cfg_attr(feature = "panic_immediate_abort", inline)]
+                #[track_caller]
+                #[rustc_const_unstable(feature = "panic_internals", issue = "none")]
+                #[lang = stringify!($lang)] // needed by codegen for panic on overflow and other `Assert` MIR terminators
+                pub const fn $lang() -> ! {
+                    // Use Arguments::new_v1 instead of format_args!("{expr}") to potentially
```
Comment (RalfJung) [inline library/core/src/panicking.rs:165]: The code doesn't use `new_v1`...

**Comment by bjorn3** [comment]
> Any way we can avoid all those mini_core copies?

The only short term alternative I know of is to use the real libcore instead, but that takes much longer to compile, makes it much harder to debug miscompilations (as reducing it to exactly the faulty part requires removing a lot more code and it is much harder to eg remove a function argument as some other code in libcore may be calling it with the original signature) and is not really suitable for bringing up a new codegen backend from scratch. The long term alternative would be to have some of those lang items be part of a virtual crate which only exists in the compiler source code. For a fair amount of trait lang items we have special code to synthesize trait implementations already. A virtual crate may do this in a cleaner way. It would also allow [rtstartup](https://github.com/rust-lang/rust/tree/master/library/rtstartup) to no longer need to define all those lang items itself twice either. And compiler-builtins may no longer need to depend on libcore depending on how many lang items are moved.

**Comment by Mark-Simulacrum** [comment]
Cherry-picked https://github.com/rust-lang/rust/pull/119440 (for the overflow checks in rustc, https://github.com/rust-lang/compiler-team/issues/724). My hope is that gives us better data in terms of the benefit here than the current set of perf.rlo benchmarks. The last [perf run](https://perf.rust-lang.org/compare.html?start=e51e98dde6a60637b6a71b8105245b629ac3fe77&end=69fcfab6febdbff1afb41a6417d0fac3a8e0abb1&stat=instructions%3Au&tab=artifact-size) there showed a +3.2MB impact to rustc.

As far as I can tell, this *should* be better; we compile to assembly roughly like this on nightly:

```
        leaq    str.0(%rip), %rdi
        leaq    .Lalloc_d6aeb8e2aa19de39a7f0e861c998af13(%rip), %rdx
        movl    $25, %esi
        callq   *_ZN4core9panicking5panic17h17cabb89c5bcc999E@GOTPCREL(%rip)
```


And like this with this PR:

```
        leaq    .Lalloc_d6aeb8e2aa19de39a7f0e861c998af13(%rip), %rdi
        callq   *_RNvNtNtCsduqIKoij8JB_4core9panicking11panic_const23panic_const_div_by_zero@GOTPCREL(%rip)
```

**Comment by Mark-Simulacrum** [comment]
> Cherry-picked #119440 (for the overflow checks in rustc, [rust-lang/compiler-team#724](https://github.com/rust-lang/compiler-team/issues/724)). My hope is that gives us better data in terms of the benefit here than the current set of perf.rlo benchmarks. The last [perf run](https://perf.rust-lang.org/compare.html?start=e51e98dde6a60637b6a71b8105245b629ac3fe77&end=69fcfab6febdbff1afb41a6417d0fac3a8e0abb1&stat=instructions%3Au&tab=artifact-size) there showed a +3.2MB impact to rustc.

OK, this did have a positive impact. The latest perf run (probably roughly comparable, we haven't changed that much code) shows a [+2.5MB](https://perf.rust-lang.org/compare.html?start=a7e4de13c1785819f4d61da41f6704ed69d5f203&end=64fbb4f0b2d621ff46d559d1e9f5ad89a8d7789b&stat=instructions%3Au&tab=artifact-size) impact. So we've cut ~0.7MB of overhead by making this change.

I think that's a pretty good demonstration that this is a good thing for us to land, so I'll go ahead and work on cleaning up the PR and see if we can do something better than ~15 lang items and duplicating messages across a bunch of surface area.

**Comment by Mark-Simulacrum** [comment]
I think this is ready for review.

r? compiler

File: compiler/rustc_middle/src/mir/terminator.rs [hunk 3]
```diff
@@ -149,44 +149,45 @@ impl<O> AssertKind<O> {
...
-    /// Get the message that is printed at runtime when this assertion fails.
+    /// Get the lang item that is invoked to print a static message when this assert fires.
     ///
     /// The caller is expected to handle `BoundsCheck` and `MisalignedPointerDereference` by
     /// invoking the appropriate lang item (panic_bounds_check/panic_misaligned_pointer_dereference)
-    /// instead of printing a static message.
-    pub fn description(&self) -> &'static str {
+    /// instead of printing a static message. Those have dynamic arguments that aren't present for
+    /// the rest of the messages here.
+    pub fn description(&self) -> LangItem {
```
Comment (saethlin) [inline compiler/rustc_middle/src/mir/terminator.rs:158]: This isn't really a description anymore, is it?

Comment (Mark-Simulacrum) [reply compiler/rustc_middle/src/mir/terminator.rs:158]: Yeah, I guess that's true. I can rename it to panic_function or something like that, happy to hear other suggestions.

**Comment by Nilstrieb** [comment]
r? Nilstrieb
Very cool! The duplication is a bit unfortunate but doesn't seem too bad. One alternative would be only having those message in the compiler and synthesizing those function when compiling libcore during expansion or whatever, but I doubt that's really better. Maybe worth a try in the future, or not. Not sure whether that's better though, probably not.
@bors r+

**Pull request merged**